File: src/commands/k8s_auth.rs
//! Kubernetes authentication analysis command.
//!
//! Analyzes Kubernetes auth method usage to understand service account
//! access patterns and identify high-volume K8s clients.
//! Supports multi-file analysis for tracking over time.
//!
//! # Usage
//!
//! ```bash
//! # Single file analysis
//! vault-audit k8s-auth audit.log
//!
//! # Multi-day analysis with CSV export
//! vault-audit k8s-auth logs/*.log --output k8s-usage.csv
//! ```
//!
//! # Output
//!
//! Displays or exports Kubernetes authentication statistics:
//! - Service account name
//! - Namespace
//! - Pod name (if available)
//! - Authentication count
//! - Associated entity ID
//!
//! Helps identify:
//! - Most active K8s service accounts
//! - Service accounts with excessive auth requests
//! - K8s authentication patterns by namespace
//! - Pods making frequent Vault requests

use crate::audit::types::AuditEntry;
use crate::utils::format::format_number;
use crate::utils::processor::{ProcessingMode, ProcessorBuilder};
use anyhow::Result;
use std::collections::HashMap;

#[derive(Debug, Clone)]
struct K8sAuthState {
    k8s_logins: usize,
    entities_seen: HashMap<String, usize>,
}

impl K8sAuthState {
    fn new() -> Self {
        Self {
            k8s_logins: 0,
            entities_seen: HashMap::with_capacity(1000),
        }
    }

    fn merge(mut self, other: Self) -> Self {
        self.k8s_logins += other.k8s_logins;
        for (entity, count) in other.entities_seen {
            *self.entities_seen.entry(entity).or_insert(0) += count;
        }
        self
    }
}

pub fn run(log_files: &[String], output: Option<&str>) -> Result<()> {
    let processor = ProcessorBuilder::new()
        .mode(ProcessingMode::Auto)
        .progress_label("Processing".to_string())
        .build();

    let (result, stats) = processor.process_files_streaming(
        log_files,
        |entry: &AuditEntry, state: &mut K8sAuthState| {
            // Filter for successful Kubernetes auth operations (response type, no error)
            if entry.entry_type != "response" || entry.error.is_some() {
                return;
            }

            let Some(request) = &entry.request else {
                return;
            };

            let path = match &request.path {
                Some(p) => p.as_str(),
                None => return,
            };

            if !path.ends_with("/login") {
                return;
            }

            // Check if it's a K8s/OpenShift login by path OR mount_type
            let is_k8s_by_path = path.contains("kubernetes") || path.contains("openshift");
            let is_k8s_by_mount = request
                .mount_type
                .as_deref()
                .is_some_and(|mt| mt == "kubernetes" || mt == "openshift");

            if is_k8s_by_path || is_k8s_by_mount {
                state.k8s_logins += 1;

                if let Some(entity_id) = entry.auth.as_ref().and_then(|a| a.entity_id.as_deref()) {
                    *state
                        .entities_seen
                        .entry(entity_id.to_string())
                        .or_insert(0) += 1;
                }
            }
        },
        K8sAuthState::merge,
        K8sAuthState::new(),
    )?;

    let total_lines = stats.total_lines;
    let k8s_logins = result.k8s_logins;
    let entities_seen = result.entities_seen;

    eprintln!(
        "\nTotal: Processed {} lines, found {} K8s logins",
        format_number(total_lines),
        format_number(k8s_logins)
    );

    println!("\n{}", "=".repeat(80));
    println!("KUBERNETES/OPENSHIFT AUTHENTICATION ANALYSIS");
    println!("{}", "=".repeat(80));

    println!("\nSummary:");
    println!("  Total lines processed: {}", format_number(total_lines));
    println!(
        "  Total K8s/OpenShift logins: {}",
        format_number(k8s_logins)
    );
    println!("  Unique entities: {}", format_number(entities_seen.len()));

    if k8s_logins > 0 {
        let ratio = k8s_logins as f64 / entities_seen.len() as f64;
        println!("  Login-to-Entity ratio: {:.2}", ratio);

        println!("\nTop 20 Entities by Login Count:");
        println!("{}", "-".repeat(80));

        let mut sorted: Vec<_> = entities_seen.iter().collect();
        sorted.sort_by(|a, b| b.1.cmp(a.1));

        for (i, (entity, count)) in sorted.iter().take(20).enumerate() {
            println!("{}. {} - {} logins", i + 1, entity, format_number(**count));
        }
    }

    if let Some(output_file) = output {
        use std::fs::File;
        use std::io::Write;
        let mut file = File::create(output_file)?;
        writeln!(file, "entity_id,login_count")?;
        for (entity, count) in &entities_seen {
            writeln!(file, "{},{}", entity, count)?;
        }
        println!("\nOutput written to: {}", output_file);
    }

    println!("\n{}", "=".repeat(80));

    Ok(())
}

File: internal/commands/k8s_auth.go
package commands

import (
	"encoding/csv"
	"fmt"
	"os"
	"sort"

	"github.com/trenner1/hashicorp-vault-audit-analysis/internal/audit"
	"github.com/trenner1/hashicorp-vault-audit-analysis/internal/processor"
	"github.com/trenner1/hashicorp-vault-audit-analysis/internal/utils"
)

// K8sAuthState tracks Kubernetes authentication statistics.
type K8sAuthState struct {
	K8sLogins    int
	EntitiesSeen map[string]int
}

// RunK8sAuth processes audit logs to extract Kubernetes authentication patterns.
//
// Analyzes audit logs for Kubernetes auth method operations, tracking service account
// authentication counts and associated entities. Outputs statistics and optionally
// writes CSV with entity login counts.
func RunK8sAuth(logFiles []string, output *string) error {
	// Create processor with default config
	cfg := processor.DefaultConfig()

	// Process files and aggregate results
	result, stats, err := processor.RunFiles(
		cfg,
		logFiles,
		func() K8sAuthState {
			return K8sAuthState{
				EntitiesSeen: make(map[string]int),
			}
		},
		processK8sAuthEntry,
		mergeK8sAuthStates,
	)

	if err != nil {
		return fmt.Errorf("process files: %w", err)
	}

	// Print stats report
	stats.Report()

	// Extract results
	totalLines := stats.TotalLines
	k8sLogins := result.K8sLogins
	entitiesSeen := result.EntitiesSeen

	fmt.Fprintf(os.Stderr, "\nTotal: Processed %s lines, found %s K8s logins\n",
		utils.FormatNumber(totalLines),
		utils.FormatNumber(k8sLogins),
	)

	fmt.Println("\n" + repeatChar("=", 80))
	fmt.Println("KUBERNETES/OPENSHIFT AUTHENTICATION ANALYSIS")
	fmt.Println(repeatChar("=", 80))

	fmt.Println("\nSummary:")
	fmt.Printf("  Total lines processed: %s\n", utils.FormatNumber(totalLines))
	fmt.Printf("  Total K8s/OpenShift logins: %s\n", utils.FormatNumber(k8sLogins))
	fmt.Printf("  Unique entities: %s\n", utils.FormatNumber(len(entitiesSeen)))

	if k8sLogins > 0 && len(entitiesSeen) > 0 {
		ratio := float64(k8sLogins) / float64(len(entitiesSeen))
		fmt.Printf("  Login-to-Entity ratio: %.2f\n", ratio)

		fmt.Println("\nTop 20 Entities by Login Count:")
		fmt.Println(repeatChar("-", 80))

		// Sort entities by count
		type entityCount struct {
			ID    string
			Count int
		}
		var sorted []entityCount
		for id, count := range entitiesSeen {
			sorted = append(sorted, entityCount{id, count})
		}
		sort.Slice(sorted, func(i, j int) bool {
			return sorted[i].Count > sorted[j].Count
		})

		for i, ec := range sorted {
			if i >= 20 {
				break
			}
			fmt.Printf("%d. %s - %s logins\n", i+1, ec.ID, utils.FormatNumber(ec.Count))
		}
	}

	// Output CSV if requested
	if output != nil && *output != "" {
		f, err := os.Create(*output)
		if err != nil {
			return fmt.Errorf("create output file: %w", err)
		}
		defer f.Close()

		w := csv.NewWriter(f)
		w.Write([]string{"entity_id", "login_count"})

		for entity, count := range entitiesSeen {
			w.Write([]string{entity, fmt.Sprintf("%d", count)})
		}

		w.Flush()
		if err := w.Error(); err != nil {
			return fmt.Errorf("write csv: %w", err)
		}

		fmt.Printf("\nOutput written to: %s\n", *output)
	}

	fmt.Println("\n" + repeatChar("=", 80))

	return nil
}

// processK8sAuthEntry processes a single audit entry for K8s auth operations.
func processK8sAuthEntry(entry *audit.AuditEntry, state *K8sAuthState) {
	// Filter for successful Kubernetes auth operations (response type, no error)
	if entry.EntryType != "response" || entry.HasError() {
		return
	}

	path := entry.Path()
	if path == "" || !pathEndsWithLogin(path) {
		return
	}

	// Check if it's a K8s/OpenShift login by path
	isK8sByPath := pathContains(path, "kubernetes") || pathContains(path, "openshift")

	// Check if it's a K8s/OpenShift login by mount type
	mountType := entry.MountType()
	isK8sByMount := mountType == "kubernetes" || mountType == "openshift"

	if !isK8sByPath && !isK8sByMount {
		return
	}

	state.K8sLogins++

	// Track entity IDs
	if entityID := entry.EntityID(); entityID != "" {
		state.EntitiesSeen[entityID]++
	}
}

// mergeK8sAuthStates combines two K8sAuthState objects.
func mergeK8sAuthStates(a, b K8sAuthState) K8sAuthState {
	a.K8sLogins += b.K8sLogins

	for entityID, count := range b.EntitiesSeen {
		a.EntitiesSeen[entityID] += count
	}

	return a
}

// Helper functions

// pathEndsWithLogin checks if a path ends with /login.
func pathEndsWithLogin(path string) bool {
	const suffix = "/login"
	return len(path) >= len(suffix) &&
		path[len(path)-len(suffix):] == suffix
}


// repeatChar repeats a character n times.
func repeatChar(c string, n int) string {
	result := ""
	for i := 0; i < n; i++ {
		result += c
	}
	return result
}
